In [2]:
import sys
sys.path.append('../')
from  mongo.lib.mongo_db  import db_conn
import pandas as pd
import numpy as np
from pyecharts import Bar,configure,Pie
import time
from es_select import ES
import itertools



In [3]:
def date_to_time(date):
    timeArray = time.strptime(date, "%Y%m%d")
    time_semp = time.mktime(timeArray)
    return time_semp   # 2013
def date_to_date(date):
    timeArray = time.strptime(date, "%a %b %d %H:%M:%S %Y")
    time_semp = time.strftime("%Y%m%d%H%M%S", timeArray);
    return time_semp   # 2013

def time_to_today(timeStamp):
    timeArray = time.localtime(int(timeStamp))
    timeArray =time.strftime("%Y-%m-%d %H:%M:%S",timeArray)
    return timeArray   # 2013

## 1、查询20181107日联通出现 5017 （尊敬的客户您好，系统检测您的访问过于频繁，为保障您的账户安全请明天再试。）

In [4]:
data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181107'),'$lte':date_to_time('20181108')},'status':5017})
data1=list(data)

In [5]:
data1_rpt_pd=pd.DataFrame(x for x in data1)

## 2、将代理信息和目标URL取出，统计请求次数、使用代理数量、代理去重列表

In [7]:
def get_proxy(data1_rpt_pd):
    all_sid_http_proxy=[]
    for sid in data1_rpt_pd.sid.values.tolist():
        es=ES('20181108')
        one_sid_http_proxy={}
        #print (sid)
        ret_list=es.get_all_ret(sid)
        #print(len(ret_list))
        one_sid_http_proxy['sid']=sid
        one_sid_http_proxy['tel']=data1_rpt_pd.loc[data1_rpt_pd.sid.values==sid,'tel'].values.tolist()[0]
        one_sid_http_proxy['status']=data1_rpt_pd.loc[data1_rpt_pd.sid.values==sid,'status'].values.tolist()[0]
        one_sid_http_proxy['http_proxy']=[]
        one_sid_http_proxy['url_base']=[]
        one_sid_http_proxy['dates']=[]
        for i,one_http in enumerate(ret_list):
            one_sid_http_proxy['http_proxy'].append(one_http['_source']['MSG']['proxies']['http'].replace('http://','').split('.')[0])
            one_sid_http_proxy['url_base'].append(one_http['_source']['MSG']['url'])
            one_sid_http_proxy['dates'].append(date_to_date(one_http['_source']['DATETIME']))
        l = [(k, len(list(g))) for k, g in itertools.groupby(one_sid_http_proxy['http_proxy'])]
        one_sid_http_proxy['http_req_count']=l
        one_sid_http_proxy['req_count']=len(ret_list)
        one_sid_http_proxy['http_unique']=list(set(one_sid_http_proxy['http_proxy']))
        one_sid_http_proxy['http_unique_count']=len(one_sid_http_proxy['http_unique'])
        all_sid_http_proxy.append(one_sid_http_proxy)
    return all_sid_http_proxy
        

In [8]:
all_sid_http_proxy=get_proxy(data1_rpt_pd)

In [9]:
all_sid_http_proxy_pd=pd.DataFrame(x for x in all_sid_http_proxy) 

In [10]:
all_sid_http_proxy_pd=all_sid_http_proxy_pd.fillna(0)

## 3、将数据按照手机号码组织，并获取单个客户的请求此次，使用的代理数量

In [11]:
def chain_l(l):
    ret_dict={}
    for one in l:
        if one[0] in ret_dict:
            ret_dict[one[0]]+=one[1]
        else:
            ret_dict[one[0]]=one[1]
    return ret_dict

In [12]:
import itertools
def get_new_tel_list(all_sid_http_proxy_pd):
    new_tel_list=[]
    for tel in all_sid_http_proxy_pd.tel.unique().tolist():
        new_tel_dict={}
        tmp=all_sid_http_proxy_pd.loc[all_sid_http_proxy_pd.tel.values==tel,['tel','req_count','http_req_count','http_unique_count','http_unique']]
        tmp=tmp.fillna(0)
        #print(tmp.shape)
        new_tel_dict['tel']=tel
        new_tel_dict['req_count']=tmp.req_count.values.sum()
        #print(tmp.http_req_count.values.tolist())
        l=list(itertools.chain(*tmp.http_req_count.values.tolist()))
        l_dict=chain_l(l)
        new_tel_dict['http_req_count']=tuple(l_dict.values())
        new_tel_dict['http_unique_count']=tmp.http_unique_count.values.sum()
        new_tel_dict['http_unique']= tuple(l_dict)
        new_tel_list.append(new_tel_dict)
    new_tel_pd=pd.DataFrame(x for x in new_tel_list)
    return new_tel_pd

In [13]:
new_tel_pd=get_new_tel_list(all_sid_http_proxy_pd)

## 4、当天只使用同一个代理但是被封的客户

In [14]:
filte=np.logical_and(new_tel_pd.http_unique_count.values==1,True)
tmp_1=new_tel_pd.loc[filte,['tel','req_count','http_unique_count','http_unique']]

In [15]:
tmp_1

tel  req_count  http_unique_count   http_unique
2   18599155300        256                  1  (squidsz80,)
4   13142010961          2                  1  (squidsz75,)
9   13140368999         28                  1  (squidsz97,)
21  15647000634         41                  1  (squidsz40,)
25  18585210620        131                  1  (squidsz73,)
26  15623253579         22                  1  (squidsz43,)
30  18659300031        140                  1  (squidsz82,)

### 电话号码不多，全部反查一下

In [16]:
data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181107'),'$lte':date_to_time('20181108')},'tel':{'$in':tmp_1.tel.values.tolist()}})
data2=list(data)
data2=pd.DataFrame(x for x in data2)
one_sid_http_proxy=get_proxy(data2)
one_tel_http_proxy_pd=pd.DataFrame(x for x in one_sid_http_proxy) 
one_tel_http_proxy_pd=get_new_tel_list(one_tel_http_proxy_pd)

In [17]:
one_tel_http_proxy_pd.loc[:,['tel','req_count','http_unique_count','http_unique']]

tel  req_count  http_unique_count  \
0  13140368999         28                  1   
1  13142010961          2                  1   
2  15623253579         22                  1   
3  15647000634        219                  4   
4  18585210620        131                  1   
5  18599155300        256                  1   
6  18659300031        140                  1   

                                    http_unique  
0                                  (squidsz97,)  
1                                  (squidsz75,)  
2                                  (squidsz43,)  
3  (squidsz27, squidsz15, squidsz40, squidsz97)  
4                                  (squidsz73,)  
5                                  (squidsz80,)  
6                                  (squidsz82,)

### 由此可以看出当天 15647000634 当天也发生了多次请求，13140368999 发生了一次请求，7日没有成功，8日成功，可能用户在其他渠道操作过，13142010961原因不明

### 接下来分析 代理squidsz75，在这次请求前后是否正常

In [18]:
es=ES('20181108')
proxy_ret=es.get_proxy_ret('http://squidsz75.dianhua.cn:8080')

In [19]:
one_sid_http_proxy_list=[]
for one in proxy_ret:
    one_sid_http_proxy={}
    one_sid_http_proxy['http_proxy']='http://squidsz75.dianhua.cn:8080'
    one_sid_http_proxy['url_base']=(one['_source']['MSG']['url'])
    one_sid_http_proxy['dates']=(date_to_date(one['_source']['DATETIME']))
    one_sid_http_proxy['sid']=((one['_source']['MSG']['sid']))
    one_sid_http_proxy_list.append(one_sid_http_proxy)


In [20]:
def sid_get_err_code(sid):
    data=db_conn.sid_info.find_one({"sid":sid,},{'_id':0,'status':1})
    return data.get('status',1)

In [21]:
one_sid_http_proxy_pd=pd.DataFrame(x for x in one_sid_http_proxy_list)
#one_sid_http_proxy_pd.groupby('sid')['http_proxy'].size().reset_index()
ip_proxy_sort_pd=one_sid_http_proxy_pd.groupby('sid')['dates'].max().reset_index().sort_values(by="dates")
ip_proxy_sort_pd['status']=ip_proxy_sort_pd.sid.map(sid_get_err_code)

In [22]:
ip_proxy_sort_pd

sid           dates  status
58  SIDd682cfe019b14165bdd5948a23836b51  20181107001621       0
31  SID4da7ba9fb2734cde93eedaedaaade412  20181107002818       0
24  SID3f60c901bcfe4d9fa0a92581e3cc1670  20181107041523    5017
10  SID16a9f0c080184319b98aa04d895207fb  20181107071637       0
60  SIDdd23743f92e4470c98aab8af4dbb687f  20181107084346       0
26  SID42807b0933d64254b3b267156c754055  20181107091936       0
55  SIDcd66979cc85d4e14b581e0b82809c36c  20181107092354    5017
25  SID40f23557b5724db88e7bea9db28804f1  20181107093852       0
51  SIDba682286991649dea22bd282f0e48989  20181107095033       0
27  SID43486c4dd4d94d99943926dc5a585e91  20181107100350       0
50  SIDb7078526da63494b8612dda3a2b8b47e  20181107101444    4999
42  SID7ce62ea1f75a491aa47ff3d1c043ac40  20181107101804       0
17  SID279da06fa50745aca00d9ef4728f88d9  20181107102856       0
62  SIDe0f9d00b73b349fe865cf7c3b3c8a0a9  20181107103920       0
38  SID6d236db909dc42a0a57d9917160f0acc  20181107104631    3003
8   SID14070bb28ca1430c869085bc6fdbf9ce  20181107105456       0
56  SIDd1740d0d3cca4243add4a3e676824408  20181107110248       0
2   SID0ac75ab77a1c4eb38d0c9257276eda05  20181107111118    3003
3   SID0afc3cb9f65945be9c496b1600ceb028  20181107111350       0
14  SID1e197b15ef344055a3c5718720af6791  20181107112133       0
6   SID118b2974569f47bdb8511d9a939443b3  20181107113354       0
45  SID9289c1fd99e0478d9a2364cf427bceb6  20181107114406       0
59  SIDd9dd7646ad33450fa39327cb462c049b  20181107115440       0
21  SID387538e5f3144d128e4b0a632ec9d7b3  20181107120135       0
19  SID2ec4fecff92a4663a2f7d4779aaeec33  20181107121034       0
0   SID00f14759c7c14e589103f99adfaf4525  20181107121545       0
54  SIDc8ca9f8d21eb420cb1f379cf9f716919  20181107123525       0
57  SIDd3d7e0026de647f19f884e1ebbc8c2cc  20181107124956       0
7   SID13bca975d3824b73b56b073936aca4d9  20181107130137    5017
63  SIDe46dcf530b95433a8cbf80a7b5d82808  20181107130807       0
..                                  ...             ...     ...
28  SID448ea54169f54c26b5d0b5c8ea3593e0  20181107142646    4107
43  SID84255e9b96054c34bd66f7832206c004  20181107143206       0
39  SID73c03e335b9d4e648f7e5f889285ae63  20181107144609       0
44  SID874b97fef13d485db7b2f2e1c6f4249c  20181107145008    5017
4   SID0c3bbc9ef68d4a4892ae423459390437  20181107145913       0
30  SID4d949baf0953483eaa66ccef44af354d  20181107151247       0
61  SIDdee1af0383184b6c974ea8eb8daecfd8  20181107151331       0
46  SIDaaacd7329e8c4819bf26af4e258d461b  20181107153234       0
36  SID61754989652d4f57aa1fc8286898f00a  20181107153352       0
40  SID76a8c863bee147fdafd937efafab90a2  20181107155347       0
37  SID64dcf7f13da545a8a625d37fbfc18000  20181107160023    5017
1   SID09c032708815458bb96aff2974954a56  20181107161131       0
13  SID1cea89cce4664cc090a8c910d42d257b  20181107163147    3003
18  SID27db7d2271f047f7a3f2df3383cc2965  20181107165054       0
11  SID179cfb38227d4170a9f1b0b2374a9fea  20181107165429       0
35  SID5f8cffc4965249708660037983908cd8  20181107171315       0
15  SID206c0f657fd44e91a47c51010437ec46  20181107172242       0
9   SID142e53ecacce47559c147ece28db7364  20181107173534    4999
29  SID467f979022ad4ca4bdfd8138cce7897d  20181107181153       0
32  SID582b535512da417da84cbe702760fc24  20181107181651       0
41  SID7b176b8483e147a391b018583b7224f4  20181107185958       0
34  SID5df50a47598f4769951d9b69d2f2529d  20181107191802    4999
53  SIDc194436dc76343dabd0354b3722a0a53  20181107194704       0
49  SIDafe78b5d37aa4339b03b4c1ca69e4743  20181107203109    5017
23  SID3a9514d0ff314405afbfa68c9fe7089a  20181107203640    4002
52  SIDbf70e96634c14e51b962cd453f1df85d  20181107211745       0
22  SID38e62709366e4aeba118d96e44597a7e  20181107214346       0
65  SIDf4118ea9611c4014a505e3b9b5b37a5e  20181107221330       0
47  SIDacbbc07904324c0f88ccaae9137bb57c  20181107234147    3003
16  SID2608c6f42d2a4166912aaa127daab89e  20181107234311       0

[66 rows x 3 columns]

#### 13142010961 当天发生的请求为 SIDafe78b5d37aa4339b03b4c1ca69e4743 ，使用代理squidsz75，请求发送在晚上20:31点，出现了请求频繁，该代理之后的其他会话中还是可以正常使用，并爬取成功，怀疑可能该客户当天通过其他渠道操作过账户。

## 5、使用多个代理被封情况分析
req_count 为请求官网次数

http_req_count 为http_unique中代理对应请求次数

http_unique_count 为使用代理个数

http_unique 为被使用过的代理

### A、使用代理数量超过1个的客户

In [23]:
filte=np.logical_and(new_tel_pd.http_unique_count.values>1,True)
tmp_2=new_tel_pd.loc[filte,['tel','req_count','http_req_count','http_unique_count','http_unique']]
tmp_2

tel  req_count                                     http_req_count  \
0   13288388003         25                  (9, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1)   
1   18666151238        240                                      (15, 15, 210)   
3   18672485919        195                           (19, 15, 4, 13, 15, 129)   
5   13250512299          4                                             (2, 2)   
6   18685794878        195                            (15, 30, 15, 4, 2, 129)   
7   15607201880        278                           (2, 15, 15, 15, 22, 209)   
8   13015415959        240                                           (239, 1)   
10  15532356119          6                                             (5, 1)   
11  18585793920        175                                  (15, 15, 15, 130)   
12  18681019079        268                                    (207, 7, 2, 52)   
13  13284253781         10                                 (5, 1, 1, 1, 1, 1)   
14  18690566658         14                        (6, 1, 1, 1, 1, 1, 1, 1, 1)   
15  13049668096         31               (7, 2, 2, 6, 4, 2, 2, 2, 1, 1, 1, 1)   
16  18695023888         12                                 (5, 3, 1, 1, 1, 1)   
17  18596783160        121                                        (2, 3, 116)   
18  13138610588         76  (2, 4, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...   
19  18138764589         10                                             (5, 5)   
20  18605979944        147                                          (16, 131)   
22  15585972487         52                                            (7, 45)   
23  13168775881        151                                      (15, 115, 21)   
24  13158265458        206              (15, 22, 6, 6, 2, 7, 2, 2, 6, 136, 2)   
27  13268712262        227                                 (7, 15, 6, 2, 197)   
28  13155152291         11                                 (6, 1, 1, 1, 1, 1)   
29  15518697667          3                                          (1, 1, 1)   
31  18691021873         14                                    (4, 2, 6, 1, 1)   
32  18688363770         59  (2, 2, 2, 2, 2, 2, 5, 2, 4, 6, 3, 2, 2, 7, 6, ...   
33  15566370820         31  (3, 4, 2, 2, 6, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
34  17695583669         12                                    (2, 3, 3, 3, 1)   
35  18686104114         18                             (2, 2, 10, 1, 1, 1, 1)   

    http_unique_count                                        http_unique  
0                  11  (squidsz26, squidsz45, squidsz81, squidsz78, s...  
1                   3                 (squidsz01, squidsz118, squidsz48)  
3                   7  (squidsz114, squidsz30, squidsz01, squidsz17, ...  
5                   2                             (squidsz73, squidsz72)  
6                   7  (squidsz98, squidsz40, squidsz23, squidsz34, s...  
7                   6  (squidsz21, squidsz15, squidsz01, squidsz75, s...  
8                   2                            (squidsz48, squidsz125)  
10                  2                             (squidsz84, squidsz85)  
11                  4       (squidsz47, squidsz72, squidsz34, squidsz28)  
12                  4       (squidsz26, squidsz20, squidsz25, squidsz33)  
13                  6  (squidsz46, squidsz92, squidsz90, squidsz75, s...  
14                  9  (squidsz76, squidsz83, squidsz85, squidsz111, ...  
15                 12  (squidsz90, squidsz118, squidsz105, squidsz01,...  
16                  6  (squidsz82, squidsz96, squidsz102, squidsz105,...  
17                  3                  (squidsz73, squidsz50, squidsz93)  
18                 36  (squidsz07, squidsz14, squidsz09, squidsz23, s...  
19                  2                            (squidsz130, squidsz26)  
20                  2                            (squidsz105, squidsz78)  
22                  2                             (squidsz07, squidsz30)  
23                  3                (squidsz124, squidsz84, squidsz105)  
24    

### B、分析代理总调用次数小于100次的客户

In [24]:
filte=np.logical_and(new_tel_pd.http_unique_count.values>1,new_tel_pd.req_count.values<100)
tmp_3=new_tel_pd.loc[filte,['tel','req_count','http_req_count','http_unique_count','http_unique']]
tmp_3

tel  req_count                                     http_req_count  \
0   13288388003         25                  (9, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1)   
5   13250512299          4                                             (2, 2)   
10  15532356119          6                                             (5, 1)   
13  13284253781         10                                 (5, 1, 1, 1, 1, 1)   
14  18690566658         14                        (6, 1, 1, 1, 1, 1, 1, 1, 1)   
15  13049668096         31               (7, 2, 2, 6, 4, 2, 2, 2, 1, 1, 1, 1)   
16  18695023888         12                                 (5, 3, 1, 1, 1, 1)   
18  13138610588         76  (2, 4, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...   
19  18138764589         10                                             (5, 5)   
22  15585972487         52                                            (7, 45)   
28  13155152291         11                                 (6, 1, 1, 1, 1, 1)   
29  15518697667          3                                          (1, 1, 1)   
31  18691021873         14                                    (4, 2, 6, 1, 1)   
32  18688363770         59  (2, 2, 2, 2, 2, 2, 5, 2, 4, 6, 3, 2, 2, 7, 6, ...   
33  15566370820         31  (3, 4, 2, 2, 6, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
34  17695583669         12                                    (2, 3, 3, 3, 1)   
35  18686104114         18                             (2, 2, 10, 1, 1, 1, 1)   

    http_unique_count                                        http_unique  
0                  11  (squidsz26, squidsz45, squidsz81, squidsz78, s...  
5                   2                             (squidsz73, squidsz72)  
10                  2                             (squidsz84, squidsz85)  
13                  6  (squidsz46, squidsz92, squidsz90, squidsz75, s...  
14                  9  (squidsz76, squidsz83, squidsz85, squidsz111, ...  
15                 12  (squidsz90, squidsz118, squidsz105, squidsz01,...  
16                  6  (squidsz82, squidsz96, squidsz102, squidsz105,...  
18                 36  (squidsz07, squidsz14, squidsz09, squidsz23, s...  
19                  2                            (squidsz130, squidsz26)  
22                  2                             (squidsz07, squidsz30)  
28                  6  (squidsz79, squidsz83, squidsz85, squidsz89, s...  
29                  3                (squidsz127, squidsz128, squidsz01)  
31                  5  (squidsz39, squidsz43, squidsz44, squidsz72, s...  
32                 20  (squidsz49, squidsz50, squidsz73, squidsz71, s...  
33                 19  (squidsz42, squidsz44, squidsz43, squidsz46, s...  
34                  5  (squidsz48, squidsz47, squidsz49, squidsz72, s...  
35                  7  (squidsz26, squidsz24, squidsz33, squidsz35, s...

#### 查看这些客户当日是否有其他返回码的请求

In [27]:
data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181107'),'$lte':date_to_time('20181108')},'tel':{'$in':tmp_3.tel.values.tolist()}})
data2=list(data)
data2=pd.DataFrame(x for x in data2)
some_sid_http_proxy=get_proxy(data2)
some_sid_http_proxy_pd=pd.DataFrame(x for x in some_sid_http_proxy) 
some_sid_http_proxy_pd=get_new_tel_list(some_sid_http_proxy_pd)

In [28]:
some_sid_http_proxy_pd

http_req_count  \
0                (7, 2, 2, 6, 4, 2, 2, 2, 1, 1, 1, 1)   
1   (2, 4, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...   
2                                  (6, 1, 1, 1, 1, 1)   
3                                              (2, 2)   
4                                  (5, 1, 1, 1, 1, 1)   
5                   (9, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1)   
6                                           (1, 1, 1)   
7                                              (5, 1)   
8   (3, 2, 4, 2, 6, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
9                                             (7, 45)   
10                                    (2, 3, 3, 3, 1)   
11                                             (5, 5)   
12                             (2, 2, 10, 1, 1, 1, 1)   
13  (2, 2, 2, 2, 2, 2, 5, 2, 4, 6, 3, 2, 2, 7, 6, ...   
14                        (6, 1, 1, 1, 1, 1, 1, 1, 1)   
15                                    (4, 2, 6, 1, 1)   
16                                 (5, 3, 1, 1, 1, 1)   

                                          http_unique  http_unique_count  \
0   (squidsz90, squidsz118, squidsz105, squidsz01,...                 12   
1   (squidsz07, squidsz14, squidsz09, squidsz23, s...                 36   
2   (squidsz79, squidsz83, squidsz85, squidsz89, s...                  6   
3                              (squidsz73, squidsz72)                  2   
4   (squidsz46, squidsz74, squidsz75, squidsz50, s...                  6   
5   (squidsz26, squidsz45, squidsz81, squidsz78, s...                 11   
6                 (squidsz127, squidsz128, squidsz01)                  3   
7                              (squidsz84, squidsz85)                  2   
8   (squidsz42, squidsz43, squidsz44, squidsz46, s...                 19   
9                              (squidsz07, squidsz30)                  2   
10  (squidsz48, squidsz47, squidsz49, squidsz72, s...                  5   
11                            (squidsz130, squidsz26)                  2   
12  (squidsz26, squidsz24, squidsz33, squidsz35, s...                  7   
13  (squidsz49, squidsz50, squidsz73, squidsz71, s...                 20   
14  (squidsz76, squidsz83, squidsz85, squidsz111, ...                  9   
15  (squidsz39, squidsz43, squidsz44, squidsz72, s...                  5   
16  (squidsz82, squidsz96, squidsz102, squidsz105,...                  6   

    req_count          tel  
0          31  13049668096  
1          76  13138610588  
2          11  13155152291  
3           4  13250512299  
4          10  13284253781  
5          25  13288388003  
6           3  15518697667  
7           6  15532356119  
8          31  15566370820  
9          52  15585972487  
10         12  17695583669  
11         10  18138764589  
12         18  18686104114  
13         59  18688363770  
14         14  18690566658  
15         14  18691021873  
16         12  18695023888

#### 这些用户当日不存在成功状态请求，由此推测可能是频繁变更IP地址引起封禁

### C、分析请求数量超过100的代理

In [29]:
filte=np.logical_and(new_tel_pd.http_unique_count.values>1,new_tel_pd.req_count.values>100)
tmp_4=new_tel_pd.loc[filte,['tel','req_count','http_req_count','http_unique_count','http_unique']]
tmp_4

tel  req_count                         http_req_count  \
1   18666151238        240                          (15, 15, 210)   
3   18672485919        195               (19, 15, 4, 13, 15, 129)   
6   18685794878        195                (15, 30, 15, 4, 2, 129)   
7   15607201880        278               (2, 15, 15, 15, 22, 209)   
8   13015415959        240                               (239, 1)   
11  18585793920        175                      (15, 15, 15, 130)   
12  18681019079        268                        (207, 7, 2, 52)   
17  18596783160        121                            (2, 3, 116)   
20  18605979944        147                              (16, 131)   
23  13168775881        151                          (15, 115, 21)   
24  13158265458        206  (15, 22, 6, 6, 2, 7, 2, 2, 6, 136, 2)   
27  13268712262        227                     (7, 15, 6, 2, 197)   

    http_unique_count                                        http_unique  
1                   3                 (squidsz01, squidsz118, squidsz48)  
3                   7  (squidsz114, squidsz30, squidsz01, squidsz17, ...  
6                   7  (squidsz98, squidsz40, squidsz23, squidsz34, s...  
7                   6  (squidsz21, squidsz15, squidsz01, squidsz75, s...  
8                   2                            (squidsz48, squidsz125)  
11                  4       (squidsz47, squidsz72, squidsz34, squidsz28)  
12                  4       (squidsz26, squidsz20, squidsz25, squidsz33)  
17                  3                  (squidsz73, squidsz50, squidsz93)  
20                  2                            (squidsz105, squidsz78)  
23                  3                (squidsz124, squidsz84, squidsz105)  
24                 11  (squidsz73, squidsz42, squidsz121, squidsz01, ...  
27                  5  (squidsz90, squidsz102, squidsz96, squidsz101,...

#### 使用代理数量相对B情况要少，但是均存在某一个代理使用超过100的情况，可能是因为当日请求量超过阈值或者Ip地址发送变化引起

### D、分析C中代理使用量超过100的情况发生的时机

In [30]:
def select_tel(tel):
    data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181107'),'$lte':date_to_time('20181108')},'tel':tel})
    data2=list(data)
    data2=pd.DataFrame(x for x in data2)
    some_sid_http_proxy=get_proxy(data2)
    some_sid_http_proxy_pd=pd.DataFrame(x for x in some_sid_http_proxy) 
    some_sid_http_proxy_pd=some_sid_http_proxy_pd.loc[some_sid_http_proxy_pd.req_count.values>0,['sid','req_count','tel','status','dates']]
    some_sid_http_proxy_pd['dates']=some_sid_http_proxy_pd.dates.map(max)
    return some_sid_http_proxy_pd

In [31]:
select_tel('18672485919')

sid  req_count          tel  status  \
0  SID5772afd8cec049b7a4c5c821a9283f3a        129  18672485919    5017   
1  SID8b29a6ec3bf448e797c2740f96cb8a72         15  18672485919    5017   
2  SID97dbe4a7add640668f87811834c06370         21  18672485919    5017   
3  SID34b0c91dc6364c56a08744fdf72deb09         15  18672485919    5017   
4  SID3c9211bc11d949e5acf419b2f64941e6         15  18672485919    5017   

            dates  
0  20181107162711  
1  20181107163442  
2  20181107164347  
3  20181107190859  
4  20181107205353

In [32]:
select_tel('15607201880')

sid  req_count          tel  status  \
0  SID1453e31456054813a3f785ed66e59956        209  15607201880    5017   
1  SID1430cd325f7a4ceb93542a5f8f73e524         22  15607201880    5017   
3  SID874b97fef13d485db7b2f2e1c6f4249c         15  15607201880    5017   
4  SIDb9ae4d5a1fdd430a8f983b953964f685         15  15607201880    5017   
5  SID1c065905cf3a43e88cebe270b33eefbd         15  15607201880    5017   
6  SID5b0679057ae04c3db63b13d629c2cab7          2  15607201880    5017   

            dates  
0  20181107144510  
1  20181107144521  
3  20181107145008  
4  20181107161726  
5  20181107161928  
6  20181107175627

#### 由此可以看出，客户均在首次请求时已经触发封禁

## 5、结论
### 1）、单次爬虫事务中，如果短时间内请求官网超过131次，可能被封禁
### 2）、多次爬虫事务中，如果使用多个ip代理请求官网，超过3个ip将可能被封禁
### 3）、联通反爬策略为，tel+ip进行识别，相同ip不同客户、相互不干扰
### 4）、针对分析结果，建立爬虫代理池对该以上反爬策略无效